lib/SILOptimizer/Utils/InstOptUtils.cpp

```diff
@@ -1865,7 +1865,7 @@ void swift::salvageDebugInfo(SILInstruction *I) {
           NewVarInfo.Type = STI->getType();
 
         // Create a new debug_value
-        SILBuilder(DbgInst, DbgInst->getDebugScope())
+        SILBuilder(STI, DbgInst->getDebugScope())
```

Comment:
I can't convince myself that it's correct to hoist a debug_value this way:

```
%s = struct %a
<other code with debug locations>, loc 1
debug_value %s, loc 2
```
==>
```
debug_value %a, loc 2
%s = struct %a
<other code with debug locations>, loc1
```

`debug_value` have specific positions relative to other debug info.

First, I'm not sure what happens to debug info when we move a `debug_value` out of order relative to other debug locations. Does the debug location of a `debug_value` instruction even matter? Does it get passed along to debug info?

Second, the "other code with debug locations" may, in theory, have a `debug_value` for the same variable. In that highly hypothetical case, the debugger could report the wrong value for the variable.

Or, maybe the first and second issue cancel each other out?
If we have this:

```
debug_value, name "x", loc 2
debug_value, name "x", loc 1
```

What does the debug information do with it?

@adrian-prantl can you share some insight?

Reply:
The SILLocation attached to a debug_value is mostly ignored, but it could be used to be a bit more principled about what we want to optimize for. 

Fundamentally debug_values will always be evaluated at a breakpoint at a real instruction below the debug_value's insertion point. Because of that it would not make sense to hoist a debug_value and insert it right above a real instruction with a "later" location. But we currently don't have a notion of ordering between SILLocations. 

I think the only rule we could actually currently implement is not never move a debug_value past another debug_value describing the same variable.